File: app/src/main/java/ru/santaev/clipboardtranslator/ui/settings/ListSettingsItem.java
package ru.santaev.clipboardtranslator.ui.settings;

import android.app.AlertDialog;
import android.content.Context;
import android.content.res.Resources;

public class ListSettingsItem extends SettingsItem {

    public interface ISettingsPropertySetter {
        void setValue(int value);
    }

    public interface ISettingsPropertyGetter {
        int getValue();
    }

    private final int textArray;
    private final int[] valueArray;
    private final String titleString;
    private String[] textArrayRes;
    private final Context context;
    private final ISettingsPropertySetter setter;
    private final ISettingsPropertyGetter getter;

    public ListSettingsItem(Context context, int title, int arrayText, int arrayValue,
                            ISettingsPropertySetter setter, ISettingsPropertyGetter getter) {
        super("", "");
        this.context = context;
        this.setter = setter;
        this.getter = getter;

        Resources resources = context.getResources();
        textArray = arrayText;
        valueArray = resources.getIntArray(arrayValue);
        textArrayRes = context.getResources().getStringArray(textArray);
        titleString = resources.getString(title);

        this.title.set(titleString);
        update();
    }

    public void onClick() {
        new AlertDialog.Builder(context)
                .setTitle(titleString)
                .setItems(textArray, (dialogInterface, i) -> {
                    this.subtitle.set(textArrayRes[i]);
                    setter.setValue(valueArray[i]);
                    dialogInterface.dismiss();
                })
                .show();
    }

    private void update() {
        String text = textArrayRes[getIdx(getter.getValue())];
        this.subtitle.set(text);
    }

    private int getIdx(int value) {
        for (int i = 0; i < valueArray.length; i++) {
            if (valueArray[i] == value) {
                return i;
            }
        }
        return 0;
    }
}

File: app/src/main/java/ru/santaev/clipboardtranslator/ui/settings/ListSettingsItem.kt
package ru.santaev.clipboardtranslator.ui.settings

import android.app.AlertDialog
import android.content.Context

typealias ISettingsPropertySetter = (Int) -> Unit
typealias ISettingsPropertyGetter = () -> Int

class ListSettingsItem(
        title: Int,
        arrayValue: Int,
        private val context: Context,
        private val textArray: Int,
        private val setter: ISettingsPropertySetter,
        private val getter: ISettingsPropertyGetter
) : SettingsItem(context.getString(title), "") {

    private val valueArray: IntArray
    private val titleString: String
    private val textArrayRes: Array<String>


    init {

        val resources = context.resources
        valueArray = resources.getIntArray(arrayValue)
        textArrayRes = context.resources.getStringArray(textArray)
        titleString = resources.getString(title)

        update()
    }

    fun onClick() {
        AlertDialog.Builder(context)
                .setTitle(titleString)
                .setItems(textArray) { dialogInterface, i ->
                    this.subtitle.set(textArrayRes[i])
                    setter.invoke(valueArray[i])
                    dialogInterface.dismiss()
                }
                .show()
    }

    private fun update() {
        val text = textArrayRes[getIdx(getter.invoke())]
        this.subtitle.set(text)
    }

    private fun getIdx(value: Int): Int {
        return valueArray.indices.firstOrNull { valueArray[it] == value } ?: 0
    }
}
